### PR TITLE
[gpt-oss] support on-device sampling

### PR DESCRIPTION
**Pull request by sraizada-tt**
## Summary

Enable on-device sampling for GPT-OSS models on Galaxy [4,8] mesh. Token sampling (top-k, top-p, temperature) runs directly on Tenstorrent devices instead of the host CPU, using the existing `SamplingGenerator` / `TTSampling` infrastructure from `models.common.sampling`.

**Key changes:**
- Integrate `SamplingGenerator` into the GPT-OSS model with automatic fallback when per-device vocab exceeds 64K
- Replace host-based `sample_host` calls with device-based sampling in the decode loop
- Pad lm_head weight to `padded_vocab_size` **before** column-parallel sharding to align device shard boundaries with TTSampling device offset strides
- Skip TP all-gather when sampling is active (TTSampling handles its own all-gather)
- Pad batch dimension to 32 (tile-aligned) for on-device sampling operations

**Bug fix (second commit):** The initial implementation padded lm_head logits *after* the column-parallel matmul. This caused TTSampling device offsets to be wrong for devices 1+:
```
Device 0 offset: 0      (correct)
Device 1 offset: 25152  (should be 25136, delta +16)
Device 7 offset: 176064 (should be 175952, delta +112)
```
Every token selected from devices 1-7 mapped to the wrong global vocab index, producing garbled output. Fixed by pre-padding the weight before sharding, matching how standard Llama models handle this.

- [x] Galaxy CI pipeline: https://github.com/tenstorrent/tt-metal/actions/runs/22307429765

### CODE REVIEW

File: models/demos/gpt_oss/tt/model.py [hunk 1]
```diff
@@ -181,6 +183,62 @@ def __init__(
...
+                mesh_device=mesh_device,
+                tt_ccl=None,
+                enable_internal_trace=False,
+            )
+            # Hook reset_sampling_params to set prefill flag — Generator calls this
+            # before prefill forward; tells _forward_layers_and_head to skip TP all-gather
+            _orig_reset = self.sampling.reset_sampling_params
+
+            def _reset_with_flag(params, _orig=_orig_reset):
+                _orig(params)
```
[the remark this answers is not in the document]
Comment (sraizada-tt) [reply models/demos/gpt_oss/tt/model.py:217]: Good point — the `_orig=_orig_reset` default-argument pattern is a common Python idiom to avoid late binding, but since `_orig_reset` is a local variable here (not a loop variable), the closure captures it correctly either way. Keeping the explicit default for clarity but either form works.

File: models/demos/gpt_oss/tt/model.py [hunk 2]
```diff
@@ -304,14 +363,19 @@ def _forward_layers_and_head(
...
-        if config.tp > 1:
+        if config.tp > 1 and not skip_gather:
             logits_gathered = self.mesh_config.allgather(
                 logits, self.ccl_manager, axis=self.mesh_config.tp_axis, dim=-1
             )
             logits.deallocate(True)
             logits = logits_gathered
+        elif skip_gather and getattr(self, "_sampling_vocab_pad", 0) > 0:
+            # Pad per-device logits to tile-aligned vocab size for TTPenalties
+            logits = ttnn.pad(logits, padding=[(0, 0), (0, 0), (0, 0), (0, self._sampling_vocab_pad)], value=0.0)
```
[the remark this answers is not in the document]
Comment (sraizada-tt) [reply models/demos/gpt_oss/tt/model.py]: Resolved in the latest commit (191cfa919e). The `_sampling_vocab_pad` attribute and the post-matmul padding are removed entirely. The lm_head weight is now padded to `padded_vocab_size` before column-parallel sharding, so each device's matmul output is already tile-aligned — no runtime padding needed.

File: models/demos/gpt_oss/demo/text_demo.py [hunk 3]
```diff
@@ -434,6 +434,15 @@ def test_gpt_oss_demo(
 
     profiler.end(f"generator_setup", iteration=batch_idx)
 
+    # Create on-device sampling params
+    SAMPLING_BATCH_SIZE = 32
+    greedy = sampling_params["temperature"] == 0
+    device_sampling_params = SamplingParams(
+        temperature=[sampling_params["temperature"]] * SAMPLING_BATCH_SIZE,
+        top_k=[1] * SAMPLING_BATCH_SIZE if greedy else [40] * SAMPLING_BATCH_SIZE,
+        top_p=[1.0] * SAMPLING_BATCH_SIZE if greedy else [sampling_params["top_p"]] * SAMPLING_BATCH_SIZE,
```
[the remark this answers is not in the document]
Comment (sraizada-tt) [reply models/demos/gpt_oss/demo/text_demo.py:473]: The hardcoded 32 is intentional and not directly related to `global_batch_size`. TTSampling requires the batch dimension to be tile-aligned (multiple of 32), and `format_sampling_params` pads the parameter lists to the device-side `max_batch_size` (always >= 32). Using `global_batch_size` here would produce shorter lists (e.g. length 1 for batch=1), which still works because `format_sampling_params` pads them — but passing 32 makes the intent explicit and matches the device batch size directly. Added a comment clarifying why.

File: models/demos/gpt_oss/tt/model.py [hunk 4]
```diff
@@ -105,7 +107,7 @@ def __init__(
         self.vocab_size = hf_config.vocab_size
         self.hf_config = hf_config
         # hf_config.num_hidden_layers = 1
-        self.core_grid = ttnn.CoreCoord(8, 8)
+        self.core_grid = mesh_device.compute_with_storage_grid_size()
```
Comment (handrewsTT) [inline models/demos/gpt_oss/tt/model.py]: Are you happy with 8x9 grid on WH-GLX?

Comment (sraizada-tt) [reply models/demos/gpt_oss/tt/model.py]: no, will revert

**Review by handrewsTT — APPROVED** [review]
Can you enable sampling for b=1 as well and check outputs before merging?
